WhateverGreen/kern_igfx.cpp

```diff
@@ -525,6 +573,58 @@ bool IGFX::wrapAcceleratorStart(IOService *that, IOService *provider) {
 	return FunctionCast(wrapAcceleratorStart, callbackIGFX->orgAcceleratorStart)(that, provider);
 }
 
+/**
+ *  ReadAUX wrapper to modify the maximum link rate valud in the DPCD buffer
+ */
+int IGFX::wrapReadAUX(void* that, void* framebuffer, uint32_t address, uint16_t length, void* buffer, void* displayPath) {
+	
+	//
+	// Abstract:
+	//
+	// Several fields in an `AppleIntelFramebuffer` instance are left zeroed because of
+	// an invalid value of maximum link rate reported by DPCD of the builtin display.
+	//
+	// One of those fields is the number of lanes is later used as a divisor during
+	// the link training, resulting in a kernel panic triggered by a divison-by-zero.
+	//
+	// DPCD are retrieved from the display via a helper function named ReadAUX().
+	// This wrapper function checks whether the driver is reading receiver capabilities
+	// from DPCD of the builtin display and then provides a custom maximum link rate value,
+	// so that we don't need to update the binary patch on each system update.
+	//
+	// If you are interested in the story behind this fix, take a look at my blog posts.
+	// Phase 1: https://www.firewolf.science/2018/10/coffee-lake-intel-uhd-graphics-630-on-macos-mojave-a-compromise-solution-to-the-kernel-panic-due-to-division-by-zero-in-the-framebuffer-driver/
+	// Phase 2: https://www.firewolf.science/2018/11/coffee-lake-intel-uhd-graphics-630-on-macos-mojave-a-nearly-ultimate-solution-to-the-kernel-panic-due-to-division-by-zero-in-the-framebuffer-driver/
+	
+	// Call the original ReadAUX() function to read from DPCD
+	int retVal = callbackIGFX->orgReadAUX(that, framebuffer, address, length, buffer, displayPath);
+	
+	// Guard: Check the DPCD register address
+	// The first 16 fields of the receiver capabilities reside at 0x0 (DPCD Register Address)
+	if (address != 0)
+		return retVal;
+	
+	// The driver tries to read the first 16 bytes from DPCD
+	// Get the current framebuffer index (field at 0x1dc in a framebuffer instance)
+	uint32_t port = *reinterpret_cast<uint32_t*>(reinterpret_cast<uint8_t*>(framebuffer) + 0x1dc);
```

Comment:
Yes, I know that using a hardcoded offset is really bad, but I am new to IOKit and not quite familiar with IORegistryEntry. 

I just checked the IORegistryExplorer. AppleIntelFramebuffer inherits from IORegistryEntry and the property "port-number" is exposed. What is a good way to cast the framebuffer instance of a type `void*` to an `IORegistryEntry*`?

Reply:
The virtual table is compatible, so you can just declare it as `IORegistryEntry*` or `IOService*` (which is closer), and it will work fine. If you want to be more certain, you could declare it `OSObject*` and do `OSDynamicCast(IOService, framebuffer)`. The latter will still not protect you from garbage pointers, so I believe there is little need for extra struggle.

Reply:
I have been playing with IORegistryEntry for a while, but unfortunately I wasn't able to retrieve the framebuffer index (yes, it is the index, AppleIntelFramebuffer@**0**, not the port number) via `WIOKit::getOSDataValue()`.   

This is what I have tried so far.

```
int wrapReadAUX(void *that, IORegistryEntry *framebuffer, ... /* extra args */)
{
    // ...
    uint32_t index = 0;
    // This invokes getOSDataValue(IORegistryEntry*, const char*, T&)
    if (WIOKit::getOSDataValue(framebuffer, "IOFBDependentIndex", index))
    {
        // Success
    }
    // ...
}
```

I have read the implementation of the [helper function](https://github.com/acidanthera/Lilu/blob/master/Lilu/Headers/kern_iokit.hpp#L37) `inline bool getOSDataValue(const OSObject *obj, const char *name, T &value);`.  

After some debugging, I found that the corresponding property `obj` (retrieved by `framebuffer->getProperty(name)`) is non-null, but `data`returned by the dynamic casting becomes NULL.  

I am not sure if I have done things incorrectly. Right now I have reverted to the hardcoded offset. I have adjusted the code style you mentioned before. Looking forward to your suggestion. Thanks for your precious time and detailed explanation.

Reply:
I managed to find some ioreg with IGPU, and discovered that it is OSNumber, not OSData. Thus you would need the following:
```C++
auto fbindexNumber = OSDynamicCast(OSNumber, framebuffer->getProperty("IOFBDependentIndex"));
if (fbindexNumber) {
  auto num = fbindexNumber->unsigned32BitValue();
}

```

Reply:
Ok, now I understand what’s going on here. Previously I thought OSData could be automatically reinterpreted as an integer, just like casting a byte buffer to an 32-bit integer in C, but these kernel types are all C++ classes.  

Sorry about this kind of stupid questions. I am new to IOKit, so I am still trying to understand these APIs, and I really appreciate your guidance.  

I will update the code later today. : )

Reply:
Resolved. I have updated the code. Now it works like a charm. : )